### PR TITLE
refactor: migrate Singpass and Corppass code to TypeScript

### PR DESCRIPTION
**Pull request by mantariksh**
This PR migrates all the backend functionality related to SingPass and CorpPass to TypeScript. It is the culmination of the following PRs: #636, #635, #656, #664, #666, #679, #667, #680, #693, #695, #609.

Only the following minor changes were made in this branch outside of those PRs:
- (chore) Log the SPCP login page HTML when login page validation fails ([59f8269](https://github.com/opengovsg/FormSG/pull/701/commits/59f82692fcd176ce6b1a2d8c848dd19d66b407c3))
- (fix) During login page validation, only fail validation if the `<title>` tags do not exist; empty strings are okay ([53c8140](https://github.com/opengovsg/FormSG/pull/701/commits/53c81409aca625dfd64f511a30411fc6db548891))

## Tests
**Email Mode SingPass**
- [x] Create a new form and add SingPass authentication with an invalid e-service ID. Attempt to activate the form. You should get an error message with a numeric error code (probably 138).
- [x] Add a valid e-service ID. Check that you can activate the form.
- [x] Add a mix of MyInfo fields which are verified (e.g. name) and non-verified (e.g. vehicle number).
- [x] Preview the form. Check that the NRIC S1234567A is automatically populated. Submit the form in preview mode and ensure that you receive the response. The response should have the [MyInfo] prefix in front of all fields, regardless of whether they were verified.
- [x] Go to the live form and log in. Check that your login has been added in the Login collection with the correct e-service ID, form ID, admin ID and agency ID.
- [x] Check that you can submit the form and receive responses.
- [x] Open your network tab and click "Log in to Singpass" on your form. Search for the "logininitial" network call, copy the URL, change the form ID in the "Target" query parameter to a non-existent form ID (or any random string), and visit the modified URL. Then log in to Singpass. You should see an empty "Not Found" page.
- [x] Click "Log in to Singpass" on your form. In a separate tab, change the authentication type of the form to "Corppass" and reactivate the form. Continue logging in to Singpass. When you are redirected back to the form, you should see a Toastr saying "There was an unexpected error with your login". Check that the `jwtSp` cookie has not been set in your browser.

**Email Mode CorpPass**
- [x] Create a new form and add CorpPass authentication with an invalid e-service ID. Check that you can activate the form (we don't do e-service ID validation for CorpPass).
- [x] Add a valid e-service ID. Check that you can activate the form.
- [x] Preview the form. Check that the UEN 123456789A is automatically populated. Submit the form in preview mode and ensure that you receive the response. You should have the UEN "123456789A" and the UID "ABC".
- [x] Go to the live form and log in. Check that your login has been added in the Login collection with the correct e-service ID, form ID, admin ID and agency ID.
- [x] Check that you can submit the form and receive responses.
- [x] Open your network tab and click "Log in to Corppass" on your form. Search for the "logininitial" network call, copy the URL, change the form ID in the "Target" query parameter to a non-existent form ID (or any random string), and visit the modified URL. Then log in to Corppass. You should see an empty "Not Found" page.
- [x] Click "Log in to Corppass" on your form. In a separate tab, change the authentication type of the form to "Singpass" and reactivate the form. Continue logging in to Corppass. When you are redirected back to the form, you should see a Toastr saying "There was an unexpected error with your login". Check that the `jwtCp` cookie has not been set in your browser.

**Storage Mode SingPass**
- [x] Create a new form and add SingPass authentication with an invalid e-service ID. Attempt to activate the form. You should get an error message with a numeric error code (probably 138).
- [x] Add a valid e-service ID. Check that you can activate the form.
- [x] Preview the form. Check that the NRIC S1234567A is automatically populated. Submit the form in preview mode.
- [x] Go to the live form and log in. Check that your login has been added in the Login collection with the correct e-service ID, form ID, admin ID and agency ID.
- [x] Check that you can submit the form and view responses.
- [x] Open your network tab and click "Log in to Singpass" on your form. Search for the "logininitial" network call, copy the URL, change the form ID in the "Target" query parameter to a non-existent form ID (or any random string), and visit the modified URL. Then log in to Singpass. You should see an empty "Not Found" page.
- [x] Click "Log in to Singpass" on your form. In a separate tab, change the authentication type of the form to "Corppass" and reactivate the form. Continue logging in to Singpass. When you are redirected back to the form, you should see a Toastr saying "There was an unexpected error with your login". Check that the `jwtSp` cookie has not been set in your browser.

**Storage Mode CorpPass**
- [x] Create a new form and add CorpPass authentication with an invalid e-service ID. Check that you can activate the form (we don't do e-service ID validation for CorpPass).
- [x] Preview the form. Check that the UEN 123456789A is automatically populated. Submit the form in preview mode.
- [x] Go to the live form and log in. Check that your login has been added in the Login collection with the correct e-service ID, form ID, admin ID and agency ID.
- [x] Check that you can submit the form and receive responses.
- [x] Open your network tab and click "Log in to Corppass" on your form. Search for the "logininitial" network call, copy the URL, change the form ID in the "Target" query parameter to a non-existent form ID (or any random string), and visit the modified URL. Then log in to Corppass. You should see an empty "Not Found" page.
- [x] Click "Log in to Corppass" on your form. In a separate tab, change the authentication type of the form to "Singpass" and reactivate the form. Continue logging in to Corppass. When you are redirected back to the form, you should see a Toastr saying "There was an unexpected error with your login". Check that the `jwtCp` cookie has not been set in your browser.

### CODE REVIEW

**Comment by mantariksh** [comment]
will complete all tests on staging before merging

File: src/app/modules/spcp/spcp.controller.ts
Comment (karrui) [inline src/app/modules/spcp/spcp.controller.ts:42]: Can we not validate this with Joi?

Comment (mantariksh) [reply src/app/modules/spcp/spcp.controller.ts:42]: lolol there is already joi validation but I added these additional checks because of this comment you left in the other PR: https://github.com/opengovsg/FormSG/pull/635#discussion_r523854401

Comment (mantariksh) [reply src/app/modules/spcp/spcp.controller.ts:42]: if you feel that Joi is sufficient then I'd also rather remove these checks

Comment (karrui) [reply src/app/modules/spcp/spcp.controller.ts:42]: aiyo i didn't know already had joi validation haha didn't connect; if got joi validation then don't need here liow, since literally only SP/CP can pass through

Comment (mantariksh) [reply src/app/modules/spcp/spcp.controller.ts:42]: eh don't liddat lol there's nothing to forgive

Comment (mantariksh) [reply src/app/modules/spcp/spcp.controller.ts:42]: but ok I shall remove the checks

**Review by karrui — APPROVED** [review]
only reviewed the last few commits that were not reviewed before

**Comment by mantariksh** [comment]
don't merge first pls, need to test corppass more thoroughly!

File: src/app/modules/form/admin-form/admin-form.controller.ts
Comment (karrui) [inline src/app/modules/form/admin-form/admin-form.controller.ts:230]: this should be in admin-form.middlewares.ts hehehe but nvm i didn't catch it before

**Pull request merged**
